runtime/lua/vim/lsp/_completion.lua

```diff
@@ -37,19 +37,49 @@ local function get_completion_word(item)
   return item.label
 end
 
----@param result lsp.CompletionList|lsp.CompletionItem[]
+--- Applies the given defaults to the completion item, modifying it in place.
+---
+--- @param item lsp.CompletionItem
+--- @param defaults lsp.ItemDefaults?
+local function apply_defaults(item, defaults)
+  if not defaults then
+    return
+  end
+
+  item.insertTextFormat = item.insertTextFormat or defaults.insertTextFormat
+  item.insertTextMode = item.insertTextMode or defaults.insertTextMode
+  item.data = item.data or defaults.data
+  if defaults.editRange then
+    local textEdit = item.textEdit or {}
+    item.textEdit = textEdit
+    textEdit.newText = textEdit.newText or item.textEditText or item.insertText
+    if defaults.editRange.start then
+      textEdit.range = textEdit.range or defaults.editRange
+    elseif defaults.editRange.insert then
+      textEdit.insert = defaults.editRange.insert
+      textEdit.replace = defaults.editRange.replace
+    end
+  end
+end
+
+---@param result vim.lsp.CompletionResult
 ---@return lsp.CompletionItem[]
 local function get_items(result)
   if result.items then
+    for _, item in ipairs(result.items) do
+      ---@diagnostic disable-next-line: param-type-mismatch
```

Comment:
This is needed because the type of `result.itemDefaults` is [anonymous](https://github.com/neovim/neovim/blob/8952a89db588db10a9dba16356f9bbd35ca5fabb/runtime/lua/vim/lsp/_meta/protocol.lua#L4776), and using an autogenerated `_anonym` type feels flaky. This is why I manually define `lsp.ItemDefaults` in `completion.lua`, which duplicates the definition :(

Happy to hear for better ideas here.

Reply:
> using an autogenerated `_anonym` type feels flaky.

We could give a better name to the generated "anon" types maybe. I don't see why we would manually define generated types, since we are generating things to avoid that.

Reply:
Okay. I don't want to block this PR waiting for the change, so I'll use the anonymous type and update it when we use a better name. WDYT @justinmk?

Reply:
See the discussion in https://github.com/neovim/neovim/pull/26724#discussion_r1436821011
Ideally we'd get rid of the anonymous types, until we have that I think it's fine to declare them manually

Reply:
So are you okay with the manual declaration done in this PR, or would you rather use the poorly named generated one (like @justinmk I believe prefers)?